hub/core/transform/transform.py

```diff
@@ -0,0 +1,148 @@
+import hub
+import math
+from typing import List
+from itertools import repeat
+from hub.core.compute.provider import ComputeProvider
+from hub.util.compute import get_compute_provider
+from hub.util.remove_cache import get_base_storage, get_dataset_with_zero_size_cache
+from hub.util.transform import (
+    check_transform_data_in,
+    check_transform_ds_out,
+    merge_all_chunk_id_encoders,
+    merge_all_tensor_metas,
+    store_data_slice,
+)
+from hub.util.exceptions import (
+    TransformComposeEmptyListError,
+    TransformComposeIncompatibleFunction,
+)
+
+
+class TransformFunction:
+    def __init__(self, func, args, kwargs):
+        """Creates a TransformFunction object that can be evaluated using .eval or used as a part of a Pipeline."""
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+
+    def eval(
+        self,
+        data_in,
+        ds_out: hub.core.dataset.Dataset,
+        num_workers: int = 1,
+        scheduler: str = "threaded",
+    ):
+        """Evaluates the TransformFunction on data_in to produce an output dataset ds_out.
+
+        Args:
+            data_in: Input passed to the transform to generate output dataset. Should support __getitem__ and __len__. Can be a Hub dataset.
+            ds_out (Dataset): The dataset object to which the transform will get written.
+                Should have all keys being generated in output already present as tensors. It's initial state should be either:-
+                - Empty i.e. all tensors have no samples. In this case all samples are added to the dataset.
+                - All tensors are populated and have sampe length. In this case new samples are appended to the dataset.
+            num_workers (int): The number of workers to use for performing the transform. Defaults to 1.
+            scheduler (str): The scheduler to be used to compute the transformation. Supported values include: 'threaded' and 'processed'. Defaults to threaded.
```

Comment:
for testing, it would be nice to also have `serial` scheduling. 

if im developing a transform, it is much faster to run entirely serial before i switch to another scheduler.

instead of adding a `serial` scheduler, supporting `num_workers=0` would also suffice